### DISCUSSION TITLE
Sprint 10

### DISCUSSION THREAD

**Discussion by Justus-at-Tazama** [tazama-lf/General-Issues#267] [General]
## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

### @cshezi - Cebolenkosi Shezi

**Yesterday:**

**Today:**

### @vorsterk - Kyle Vorster

**Yesterday:**

**Today:**

### @Lenbkr - Len Bekker

**Yesterday:**

**Today:**

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:**

**Today:**

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:**

### @johanfol - Johan Foley

**Yesterday:**

**Today:**

**Comment by Justus-at-Tazama** [comment]
## 2023/09/11

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Slides and demo video. Started working on verification of the APM variables across the platform. Catch-up of the missing transactions from Nitro.

**Today:** Complete APM variable verification. Complete missing transaction ingestion into Nitro.

### @vorsterk - Kyle Vorster

**Yesterday:** On leave on Friday. On Thursday ran performance benchmarks. Testing stringify branch but achieved similar results. Discovered a missing index in one of the rules. During the disabling of pain.001 we flipped variable focus to an unindexed field. Achieved 500TPS on only 9 big-boy nodes. Performed a few more runs when the network was quiet. Started chasing the high-score, but was unable to achieve any new results - suspect the current infrastructure mix is not appropriate for a higher scale.

**Today:** Chasing a new high score. Compile a report for the sprint demo this evening. Report on our current journey to the 500TPS at lower cost.

### @Lenbkr - Len Bekker

**Yesterday:** Assisted JP with library code to complete unit tests and reviewed the PR. Split DB approved. Looked at optimising the library for the builder functions. Can perhaps look at velocipack in the future. Came to some conclusions about options. No specific performance improvements on the first pass. Worked to make the library more robust.

**Today:** Demo and presentation. Will pick up the Docker compose for a local full-stack deployment.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Completed the splitting of the transactions collection into its own DB. Deployed to SB1 as "on". Default deployment is "off". (@rtkay123 - Please deploy to SB4 for QA.) Consider for a performance benchmark differential - Foley suggested including this in the high-score.

**Today:** Combine TADProc/CADProc. Eliminate redundant / duplicated code.

### @rtkay123 - Rodney Kanjala

**Yesterday:**

**Today:** Getting his VISA

### @johanfol - Johan Foley

**Yesterday:** Design for protobuffer development.

**Today:** Complete protobuffer design. Working on report and the 4 ISO messages. Create a master "class". PoC Reporting.

## Parking lot

Performance graph

**Comment by Justus-at-Tazama** [comment]
## 2023/09/12

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Patch batch ingestion. Synching APM logging variable names across the platform. Research on Flame-graph tools. Under active development.

**Today:** Needs a new ticket.

### @vorsterk - Kyle Vorster

**Yesterday:** Demo and presentation prep. Performance benchmarks. 1070 TPS - Redis is sus. Intermittent deployment-induced issues with E2E. Kyle suffered performance failure. Scaled up Redis and NATS. APM logging disablement wasn't pulled into Main properly.

**Today:** Chase the elusive 3000! Go Kyle! You can do it!

### @Lenbkr - Len Bekker

**Yesterday:** Demo video and presentation. Docker compose to allow local full stack deployment. This is apparently a lot of fun. @johanfol - can you do something to make this less fun? Set up to create databases and collections and populate the configurations. Working out how to deploy indexes automatically.

**Today:** Carry on carrying on. Reconnect feature design.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Worked on merging CADProc and TADProc. Looking for tests.

**Today:** Switch on Split DB to SB1 for QA. Then deploy CADProc/TADProc merge to SB1 for QA. Test and demo CATADProc. Send PR to Foley as well.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Visa, las vegas! Opened a logging ticket for the levels that were not configured. Addressed Foley's comments on the PR. Improve how the typology processor works to serialise once instead of in each iteration.

**Today:** Rename the pacs.008 field? No, let's find Rodney something else to do.

### @johanfol - Johan Foley

**Yesterday:** Updated PoC Reports. Submitted for review. Protobuffer classes and messages complete. Tests with sample JSON objects complete. We can start with ProtoBuffer tickets.

**Today:** Hand out juicy work. Review report review feedback for review. Review indexes to ensure application of indexes are automated during deployment.

## Parking lot

CATADProc.

Work for Rodney: Protobuffers

Work for Shezi: Protobuffers

**Comment by Justus-at-Tazama** [comment]
## 2023/09/13

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Started working on the implementation of ProtoBuff in the library. Rodney assisted with a hiccup. Missing rule and typology results from the protobuff file - now added. Keeping work on a single branch until tested. Implemented the encoding part and appears to be working.

**Today:** Complete the decoding side. Refine the implementation as we work with it.

### @vorsterk - Kyle Vorster

**Yesterday:** Performance benchmarking. INfrastructure chase showed a bottleneck in Redis. Scaled up Redis to 36 pods and NATS to 12. Split Redis over multiple instances. Remixed the infrastructure. Scaled up TADProc and the rule processors. Most rule processors were running at 600ms. Some rules (Mirroring, e.g.) were running at about 5s. Scaled up these processors and achieved 120ms. TMS ingestion rate achieved: 2500+ MPS!! Arango currently running at 60% memory/CPU.

**Today:** Trying to ramp up the FTPS! May re-test the Rust typology processor as well.

### @Lenbkr - Len Bekker

**Yesterday:** Full-stack local Docker compose. Populated remaining config collection tables. Based on current dev environment. Got everything running except for the rules. Will start with rule 901 as the only public rule. Private rules are a bit of a challenge.

**Today:** Cross-user testing. Health-check on Arango to check the configs and indexes are deployed.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Testing transactions collection split and updates (push to SAB4 with default setting (OFF)). Testing the CATADProc. Testing some business use cases. Found a missed transaction, but currently unable to replicate - keep an eye out for this, y'all.

**Today:** Complete unit tests for CATADProc. Integrate feedback. New ticket today.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Started working on implementation of ProtoBuff into the redis. Investigated storage options and patterns. Supported Shezi with the implementation of ProtoBuff into the library.

**Today:** Assist Len with Docker compose testing. Continue with redis/ProtoBuff implementation.

### @johanfol - Johan Foley

**Yesterday:** PoC Reporting. KPI-related meetings. Design Authority. Reviewed the indexes on the dev environment.

**Today:** Create the queries to create the indexes. Coordinate with Len to implement.

## Parking lot

Demo of CATADproc. Channel processing will now and forever more be performed in the TADProc. The location of this work is not optional. RIP the CADProc. Move documentation into the TADProc - Archive CADProc documentation.

Protobuff optional fields

Time-relative rule queries

**Comment by Justus-at-Tazama** [comment]
## 2023/09/14

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Deployed Protobuff to SB4.

**Today:** Promote changes to main - versioning is a knife's edge.

### @vorsterk - Kyle Vorster

**Yesterday:** Ran 12 performance benchmarking tests. Some runs lost transactions. Still peaking at 1200. TMS API ingestion rate is increasing. Downstream processor scaling seems to drive increases in the TMS API. Reached 2700 on ingestion. Max FTPS 1294. Added another VM for Redis. Ingesting 50% more transactions than our previous high-score.

**Today:** Turned on APM to see the bottleneck in the TADProc. Added another VM for NATS to scale up from 9 to 12 PODS. Error rate of 600 over 95k. This morning FTPS at 1388. With APM running on TADProc. Will review TADProc to identify the bottlenecks.

### @Lenbkr - Len Bekker

**Yesterday:** Tested FSDActio on JP's machine from a clean slate. Worked! Updated Arango scripts. Updated indexes. Got rule 901 to work to produce an E2E report. FSDActio takes 3 minutes to spin up! FSDActio is parameterisable with any specific GitHub branch.

**Today:** Documentation. 2 main use cases: developer support for specific work on the platform. Shareability. Will push a public and private version of FSDActio to accommodate public and private rules. Automated tested on deploy to make sure the E2E is working to start with.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Unit tests for (CA)TADProc. Tests were not reusable. Completing tests for every use case and decision branch. Tested Full-Stack Docker, works great!

**Today:** Complete testing, new ticket.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Integrating the writing of binary to redis instead of the full JSON. Need to update the library - still in the branch WIP. 

**Today:** Test E2E on SB4 as soon as Shezi is done with his change. Binary reading across all the processors.

### @johanfol - Johan Foley

**Yesterday:** Updated reporting for the PoC. Justus to review.

**Today:** TADProc performance debugging. Maybe start looking at Bun in preparation for DA discussion.

## Parking lot

Performance testing on redis protobuff implementation: in general, devise a defensible performance test that illustrates performance before and after the test. Not necessary in this case since we will be testing the impact of the ProtoBuff implementation specifically in the performance benchmarking. If in doubt, call on a higher power.

**Comment by Justus-at-Tazama** [comment]
## 2023/09/15

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Cleaned up SB4 and deployed ProtoBuff. (NATS REST Proxy coming up). Recommended to redeploy all processors.

**Today:** May start one of the redis tickets today.

### @vorsterk - Kyle Vorster

**Yesterday:** Performance benchmarking. Maxed out at 1500 FTPS. New high score! Slightly higher hardware allocation. Aiming to work on the right balance of the distribution of infrastructure. Reduced TMS API from 65 to 60 pods and was still able to achieve the ingestion rate of 2500 TPS - there appears to be some headroom in the TMS API allocation. Some problems during startup and awareness of the situation without actual logging enabled. 

**Today:** Testing E2E Protobuff for interservices communication. Justus to issue an updated configuration today to exclude redundant typologies.

### @Lenbkr - Len Bekker

**Yesterday:** Docker compose FSDActio! There are ways to trigger a post-payload through scripting. Timing of the deployment of the stack is a but random so automating the sense-check is unpredictable and not necessarily viable. Assisted Greg with a local deployment. Updated the readme with some learnings. Health checks in Docker compose are fairly sad and useless. Removed for now.

**Today:** Review the issue where the application is in a dead state if redis is an issue.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Wrapping up the testing of the TADProc. Running on SB4 and awaiting QA. Started working on the ticket for reconnecting of stateful services in NATS. Reconnection is automated, but will log the connection as it happens. Please share the Redis ticket with JP once written.

**Today:** Complete reconnection. Then grab a new ticket.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Worked on the startup library to add redis changes for the Protobuff stuff. PR open for review. Expecting to merge to main shortly. Something wonky.

**Today:** Merge the library and then update the processors that rely on redis cache interaction.

### @johanfol - Johan Foley

**Yesterday:** PoC Reporting complete. Performance benchmarking support. Justus to share reporting with the client.

**Today:** Get a drop-in replacement of Bun going to test vs Node.js.

## Parking lot

Rodney's last thing: Optimised Rusty a bit. Appears to be _blazingly fast_ - 3 to 4 times faster than Node. Document test methodology for sharing with the other processor-based experiments currently in progress (GoLang, Bun, etc).

Redis issue: Document problem in the ticket and share with JP.

Adding metadata about the state of the software at the time of a commit to a commit.

**Comment by Justus-at-Tazama** [comment]
## 2023/09/18

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Completed the updates to the NATS REST Proxy for ProtoBuff integration. Started implementing ProtoBuff for Redis out of the typology processor. Struggled a but with the set() method and worked with Rodney to find a generic way. A generic method did not appear to offer any savings. Found a way to save half of the object size.

**Today:** Submit PR for review and review. Assisting Rodney with the remaining processors.

### @vorsterk - Kyle Vorster

**Yesterday:** Performance benchmarking. 19 different tests mostly focused on infrastructure remix. ProtoBuff made a phenomenal change to the platform. Updated the config (two fewer typologies appeared to make a really big difference). Reduced CPU load as well Arango idling at 60% CPU. Processors were running at 65 to 75%. FTSP achieved: 1925 (NEW HIGH SCORE!!)

**Today:** 2000+ FTPS today! ($1000 (3y reserved) environment tests to start next week). ProtoBuff for TMS/Redis merged - will also test today.

### @Lenbkr - Len Bekker

**Yesterday:** Investigating the issue where Redis appears frozen and how to recover if the console isn't logging. Resolved for post-startup; if pre-startup, we'll need a deeper investigation. Set a 10 second time-out. Needs more detail from the issue background.

**Today:** More detailed investigation and consultation with Kyle to wrap up the ticket. Then a new ticket.

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Reconnecting to stateful services. Started Rule 001 time-frame ticket. Completed, to be tested.

**Today:** CADProc to be merged today. Testing rule 001. New ticket.

### @rtkay123 - Rodney Kanjala

**Yesterday:** WSL died. Worked around it with a new environment. Investigated a way to use ProtoBuff with Redis. TMS merged.

**Today:** Tidying up the rest of the processors today. Started with the Jupyter deployment.

### @johanfol - Johan Foley

**Yesterday:** Performance benchmarking.

**Today:** Bun!

## Parking lot

Cost-based testing: Ranges in $1000, $5000 and $10000 and performance achievable.

**Comment by JeanPierreNell** [comment]
## 2023/09/19

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
None

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Went on a call with Rodney, worked on Redis buffer stuff. Worked on it and tried to make sure everything was optimized. Bytes were decreased. Discovered a bug on TP where it is communicating with CMS.

**Today:** Looking at improving Protopuff. Pick up new ticket. 

### @vorsterk - Kyle Vorster

**Yesterday:** Worked on performance testing on the protobuff redis changes (Ran 13 Tests) , the max FTPS reached was 1866 . For some reason it didn't want to increase above that number , I scaled up redis , the rules and processor pods , increased the threads and still nothing. 

**Today:** Working on the infrastructure documentation and step for the $1000 , $5000 and $10000 environments.

### @Lenbkr - Len Bekker

**Yesterday:** Jp, Len and kyle was in short call to determine issue with Zombie state for some pods. Without logs you wouldn’t know. Main Catch block log, but don’t break. Wrote a ticket with Justus. Needs approval from Foley. Worked on ticket related to make it more robust. Polish retry strategies. Started with Rule 002.

**Today:** Continue with Rule 002 , then starting next rule. 

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Refactored Protopuff changes into cadaproc, worked on testing for rule 001 upgrades. Sat with Len and Kyle to determine issue with reconnection to stateful services. Created Ticket.

**Today:** Continue with Rule upgrades , merge cadaproc.

### @rtkay123 - Rodney Kanjala

**Yesterday:** Helped to integrate Redis serialization. Made some testing to make sure everything is above board. Didn’t get to start on Jupiter just yet. 

**Today:** Continue with Jupiter Notebook. Doubling back with Foley in case of changes.

### @johanfol - Johan Foley

**Yesterday:** Bun , drop in replacement is interesting term, quite a bit involved. Got it to start building, bit of a process, almost done.

**Today:** Continue with Bun today. 

## Parking lot

Go TP

**Comment by vorsterk** [comment]
**### 2023/09/20**
**### Attendees**
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

Apologies
None

**### Minutes:**
@cshezi - Cebolenkosi Shezi
Yesterday: Meeting with protobuff sterilization . TADProc fix , add additional unit tests added, Opened PR . Cleaned up typology processor - variables renamed **heads up**. Discovered a bug on TP where it is communicating with CMS.

Today: Rules ticket - time relative queries open rule tickets - TODO.

@vorsterk - Kyle Vorster
Yesterday: Worked on documentation for the cost of different reserve pricings for 1000 , 5000 and 10000 . 

Today: Wrap up the documentation for the infrastructure specs for the reserve pricings and will see how far I get today , and will run a benchmark on a Sandbox.

@Lenbkr - Len Bekker
Yesterday: Session regarding protobuff sterilization , which created additional tasks that took a  for chunk of the day , rule002 had a few blockers. Testing soon. Multiple threads of TADP and CRSP task to be created.

Today: Work on rule002 and wrap that up , next task: exit for NATs redis that will fix the disconnect issue - after rule002

@JeanPierreNell - Jean-Pierre Nell
Yesterday: Meeting for protobuff sterilization  , continued rule 001 , 003, 004 and 006 , in test and continue rule 007 .

Today: Finishing off the rule tickets.

@rtkay123 - Rodney Kanjala
Yesterday: Meeting for protobuff sterilization , Understanding how Jupiter works , google note book as review on work . One to one copy paste. 

Today: Continue with Jupiter Notebook. Pick task for Jupiter work require as per Justus.

@johanfol - Johan Foley
Yesterday: Meeting for protobuff sterilization and rewriting TP tests , lots of meetings.

Today: Continue with Bun today.

**### Parking lot**
Ticket for the CRSP and TADP - Len will create the ticket
Rodney ticket assigned - Time processing for Jupiter

**Comment by rtkay123** [comment]
## 2023/09/21

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@rtkay123 - Rodney Kanjala

## Apologies
Justus @Justus-Ortlepp-frms 

## Minutes:

### @cshezi - Cebolenkosi Shezi

**Yesterday:** Working with @JeanPierreNell to make rules time relative. Postman not working as per Justus' testing collection. There was one test scenario left, some struggles to get it to pass. Found a solution, thanks to Justus, missing variable in env. Everything passing but one rule. Completed 7

**Today:** Continuing with user stories

### @vorsterk - Kyle Vorster

**Yesterday:** Working on docs for 1000, 5000, 10000 reserve pricing for system. 

**Today:**
Almost done, diagrams are pending, adding scaling and monitoring to docs (blank for now). Possible runs today. 

### @Lenbkr - Len Bekker

**Yesterday:** Focused on rule 2, deployed to Sandbox 1. Ticket is on QA, however, Postman collections were not working. After receiving a collection that works, there was some retesting. Worked on "CRSP Max Cpu" issue, helped JP and Shezi. Started "Process Exit" issue. 


**Today:** Continuing "Process Exit" issue

### @JeanPierreNell - Jean-Pierre Nell

**Yesterday:** Worked on rules, ran against Justus' collections. Couple of issues on some rules, 8 rules cleared. Waiting on Justus on whether to take on another rule.

**Today:** Sybrin stuff

### @rtkay123 - Rodney Kanjala

**Yesterday:** Reading Arango python sdk docs. Issues connecting Jupyter's container to local arango container. Had to use container's IP address. Asked for more clarity on ticket, along with acceptance criteria. Got as far as generating graphs

**Today:** Generating more graphs

### @johanfol - Johan Foley

**Yesterday:** Bun stuff, build, run, connect to nats, ubuntu but tests need revision, currently broken.

**Today:** Working on aforementioned tests

## Parking lot
